IntelliTect.TestTools.TestFramework/IntelliTect.TestTools.TestFramework/TestBuilder.cs

```diff
@@ -65,147 +83,89 @@ public TestBuilder AddDependencyService<T>()
         /// <returns>This</returns>
         public TestBuilder AddDependencyInstance(object objToAdd)
         {
+            // Need to add some testing around this to see if it behaves in a similarly odd fashion as AddLogger when running tests in parallel
             Services.AddSingleton(objToAdd.GetType(), objToAdd);
             return this;
         }
 
+        // Are there other cases where we'll need to add something at this level?
+        // If so, this shouldn't be called "AddLogger".
+        // Might need to make this scoped. It's behaving oddly when running tests in parallel
+        // But only on the "Starting test case" call
+        public TestBuilder AddLogger<T>() where T : ILogger
+        {
+            Services.AddSingleton(typeof(ILogger), typeof(T));
+            return this;
+        }
+
         public void ExecuteTestCase()
         {
             #region move to a Build() method and validate all dependencies are satisfied?
             var serviceProvider = Services.BuildServiceProvider();
             #endregion
 
-            var logger = serviceProvider.GetService<ILogger>() ?? new Log();
-            Exception testBlockException = null;
-            using (var scope = serviceProvider.CreateScope())
+            
+            //Exception testBlockException = null;
+            using (var testCaseScope = serviceProvider.CreateScope())
             {
-                HashSet<object> testBlockResults = new HashSet<object>();
-                foreach (var tb in TestBlocksAndParams)
-                {
-                    logger.Info($"---{tb.TestBlockType}---");
-                    logger.Debug($"Fetching dependencies.");
-                    object testBlockInstance = null;
+                var logger = testCaseScope.ServiceProvider.GetService<ILogger>() ?? new Log();
```

Comment:
Usually what you do is register this as the default logger in your container, and then let callers override it. Hard coding the `new Log()` means it is not easy for someone to remove logging if they wanted to.

Reply:
Just to make sure I understand, will the subsequent logger calls all then be null checked (logger?.Info("blah blah blah");) to simply not log at all should the person using the package do that?

Reply:
Yea you would then null check usages of `logger` to make sure they don't cause an exception. If you want the default implementation of `ILogger` to be a `Log` you would set that up in your `Services` pre-emptively. Then in "AddLogger" you would replace that instance.

Reply:
I couldn't find a way to call the AddLogger method without enabling preview features (if the new nullable reference types features works that way for generic type constraints...) so I added a RemoveLogger method for now. Thoughts on that approach until C# 8.0 is out of preview?

Reply:
@Keboo I made some additional changes to the ILogger interface... do you want me to abandon this PR and put in a new one? What's going to be easiest for you to review?

Reply:
Also had to add an additional method to override the test case name/key. xUnit uses the constructor as test setup, so when I instantiated the TestBuilder there, the key was ".ctor". Instead of trying to guess the test case name based on that, I added a method that uses the same [CallerMemberNameAttribute] that a person can call from their test. It's convoluted, and maybe you have a better idea, but it seemed the most straightforward to handle that I don't necessarily know how someone might do test setup/teardown.